packages/dds/tree/src/core/tree/anchorSet.ts

```diff
@@ -604,16 +642,38 @@ export class AnchorSet implements ISubscribable<AnchorSetRootEvents> {
 				this.moveChildren(count, srcPath, { parent, parentField, parentIndex: start });
 			},
 			onSetValue: (value: Value): void => {
-				maybeWithNode((p) => p.events.emit("valueChanging", p, value));
+				maybeWithNode((p) => {
+					p.events.emit("valueChanging", p, value);
+				});
+				assert(parent !== undefined, "Must be in a node to set its value");
+				for (const visitors of pathVisitors.values()) {
+					for (const pathVisitor of visitors) {
+						pathVisitor.onSetValue(parent, value);
+					}
+				}
 			},
 			enterNode: (index: number): void => {
 				assert(parentField !== undefined, 0x3ab /* Must be in a field to enter node */);
 				parent = { parent, parentField, parentIndex: index };
 				parentField = undefined;
-				maybeWithNode((p) => p.events.emit("subtreeChanging", p));
+				maybeWithNode((p) => {
+					// avoid multiple pass side-effects
+					if (!pathVisitors.has(p)) {
+						const visitors: PathVisitor[] = p.events.emitAndCollect(
+							"visitSubtreeChanging",
+							p,
+						);
+						if (visitors.length > 0) pathVisitors.set(p, visitors);
+					}
+					p.events.emit("subtreeChanging", p);
+				});
 			},
 			exitNode: (index: number): void => {
 				assert(parent !== undefined, 0x3ac /* Must have parent node */);
+				maybeWithNode((p) => {
+					// check for clarity
+					if (pathVisitors.has(p)) pathVisitors.delete(p);
```

Comment:
I'd prefer the faster code, which the comment explaining what its doing:
```suggestion
					// Remove subtree path visitors added at this node if there are any
					pathVisitors.delete(p);
```